File: Mobile/lib/models/custom_request.dart
import 'user.dart';

class CustomRequestModel {
  CustomRequestModel({
    required this.id,
    required this.description,
    required this.offeredPrice,
    this.counterPrice,
    this.status = 'requested',
    this.paymentStatus = 'unpaid',
    this.fanId,
    this.creatorId,
    this.deliveryContentId,
    this.dueAt,
    this.createdAt,
    this.fan,
    this.creator,
  });

  final String id;
  final String description;
  final double offeredPrice;
  final double? counterPrice;
  final String status;
  final String paymentStatus;
  final String? fanId;
  final String? creatorId;
  final String? deliveryContentId;
  final DateTime? dueAt;
  final DateTime? createdAt;
  final UserModel? fan;
  final UserModel? creator;

  factory CustomRequestModel.fromJson(Map<String, dynamic> json) {
    return CustomRequestModel(
      id: json['id']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      offeredPrice: double.tryParse(json['offeredPrice']?.toString() ?? '') ?? 0,
      counterPrice: json['counterPrice'] != null
          ? double.tryParse(json['counterPrice'].toString())
          : null,
      status: json['status']?.toString() ?? 'requested',
      paymentStatus: json['paymentStatus']?.toString() ?? 'unpaid',
      fanId: json['fanId']?.toString(),
      creatorId: json['creatorId']?.toString(),
      deliveryContentId: json['deliveryContentId']?.toString(),
      dueAt: DateTime.tryParse(json['dueAt']?.toString() ?? ''),
      createdAt: DateTime.tryParse(json['createdAt']?.toString() ?? ''),
      fan: json['fan'] != null
          ? UserModel.fromJson(Map<String, dynamic>.from(json['fan'] as Map))
          : null,
      creator: json['creator'] != null
          ? UserModel.fromJson(Map<String, dynamic>.from(json['creator'] as Map))
          : null,
    );
  }
}

File: Backend/models/CustomRequest.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');

const CustomRequest = sequelize.define('CustomRequest', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  fanId: {
    type: DataTypes.UUID,
    allowNull: false,
    field: 'fan_id',
    references: { model: 'users', key: 'id' },
  },
  creatorId: {
    type: DataTypes.UUID,
    allowNull: false,
    field: 'creator_id',
    references: { model: 'users', key: 'id' },
  },
  description: {
    type: DataTypes.TEXT,
    allowNull: false,
  },
  offeredPrice: {
    type: DataTypes.DECIMAL(10, 2),
    allowNull: false,
    field: 'offered_price',
  },
  counterPrice: {
    type: DataTypes.DECIMAL(10, 2),
    field: 'counter_price',
  },
  status: {
    type: DataTypes.ENUM('requested', 'countered', 'accepted', 'declined', 'delivered', 'cancelled'),
    defaultValue: 'requested',
  },
  deliveryContentId: {
    type: DataTypes.UUID,
    field: 'delivery_content_id',
    references: { model: 'content', key: 'id' },
  },
  dueAt: {
    type: DataTypes.DATE,
    field: 'due_at',
  },
  paymentStatus: {
    type: DataTypes.ENUM('unpaid', 'paid', 'refunded'),
    defaultValue: 'unpaid',
    field: 'payment_status',
  },
  paidAt: {
    type: DataTypes.DATE,
    field: 'paid_at',
  },
}, {
  tableName: 'custom_requests',
  underscored: true,
  indexes: [
    { fields: ['fan_id', 'status'] },
    { fields: ['creator_id', 'status'] },
  ],
});

module.exports = CustomRequest;
